core/src/main/java/io/questdb/griffin/engine/join/MarkoutHorizonRecordCursorFactory.java

```diff
@@ -0,0 +1,631 @@
+        this.masterColumnIndex = masterColumnIndex;
+        this.slaveColumnIndex = slaveColumnIndex;
+
+        RecordArray slaveRecordArray = null;
+        try {
+            slaveRecordArray = new RecordArray(
+                    slaveFactory.getMetadata(),
+                    slaveRecordSink,
+                    configuration.getSqlHashJoinValuePageSize(),
+                    configuration.getSqlHashJoinValueMaxPages()
+            );
+            this.cursor = new MarkoutHorizonRecordCursor(
+                    columnSplit,
+                    masterColumnIndex,
+                    slaveColumnIndex,
+                    slaveRecordArray
+            );
+        } catch (Throwable th) {
+            Misc.free(slaveRecordArray);
+            throw th;
+        }
+    }
+
+    @Override
+    public boolean followedOrderByAdvice() {
+        // We handle the ORDER BY internally by emitting rows in sorted order
+        return true;
+    }
+
+    @Override
+    public RecordCursor getCursor(SqlExecutionContext executionContext) throws SqlException {
+        RecordCursor masterCursor = masterFactory.getCursor(executionContext);
+        RecordCursor slaveCursor = null;
+        try {
+            slaveCursor = slaveFactory.getCursor(executionContext);
+            cursor.of(masterCursor, slaveCursor, executionContext.getCircuitBreaker());
+            return cursor;
+        } catch (Throwable ex) {
+            Misc.free(masterCursor);
+            Misc.free(slaveCursor);
+            throw ex;
+        }
+    }
+
+    @Override
+    public int getScanDirection() {
+        return masterFactory.getScanDirection();
+    }
+
+    @Override
+    public boolean recordCursorSupportsRandomAccess() {
+        return false;
+    }
+
+    @Override
+    public boolean supportsUpdateRowId(TableToken tableToken) {
+        return masterFactory.supportsUpdateRowId(tableToken);
+    }
+
+    @Override
+    public void toPlan(PlanSink sink) {
+        sink.type("Markout Horizon Join");
+        sink.meta("timestampColumn").val(masterColumnIndex);
+        sink.meta("offsetColumn").val(slaveColumnIndex);
+        sink.child(masterFactory);
+        sink.child(slaveFactory);
+    }
+
+    @Override
+    protected void _close() {
+        Misc.freeIfCloseable(getMetadata());
+        Misc.free(masterFactory);
+        Misc.free(slaveFactory);
+    }
+
+    /**
+     * Record cursor that emits cross-join results in timestamp order.
+     * <p>
+     * We materialize the slave cursor (arithmetic sequence) into a RecordArray
+     * because we need random access to its rows.
+     */
+    private static class MarkoutHorizonRecordCursor extends AbstractJoinCursor {
+        // Native memory layout constants
+        private static final int BLOCK_HEADER_SIZE = 16;
+        private static final int BLOCK_OFFSET_NEXT_BLOCK_ADDR = 0;    // long (8 bytes)
+        private static final int BLOCK_OFFSET_NEXT_FREE_SLOT = 8;     // int (4 bytes)
```

Comment:
nit: can we optimize memory usage here? For instance, `BLOCK_OFFSET_USED_SLOT_COUNT` and `BLOCK_OFFSET_USED_SLOT_COUNT` fields could be `short`s. One more simple step is moving all blocks to a heap (grow-only off-heap buffer) and align them to int/long. This way we can use compressed pointers in `ITERATOR_OFFSET_NEXT_ITER_ADDR` (`int` instead of `long`).

Optionally, we could go one step further and check the number of right-hand rows somewhere in `detectMarkoutHorizonPattern()` and validate that it fits `short`. This way, `ITERATOR_OFFSET_OFFSET_FROM_BLOCK_START` could be change to index instead of offset (`short` instead of `int`) and `ITERATOR_OFFSET_NEXT_SLAVE_ROW_NUM` can also be `short`.

Reply:
Block header is 16 bytes vs. block size of 40 kilobytes, so shrinking `usedSlotcount` and `nextFreeSlot` won't move the needle.